streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -461,120 +461,125 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
         // have existed with the expected number of partitions, or some create topic returns fatal errors.
         log.debug("Starting to validate internal topics {} in partition assignor.", topics);
 
-        long currentWallClockMs = time.milliseconds();
+        final long currentWallClockMs = time.milliseconds();
         final long deadlineMs = currentWallClockMs + retryTimeoutMs;
 
-        Set<String> topicsNotReady = new HashSet<>(topics.keySet());
+        final Set<String> topicsNotReady = new HashSet<>(topics.keySet());
         final Set<String> newlyCreatedTopics = new HashSet<>();
 
         while (!topicsNotReady.isEmpty()) {
-            final Set<String> tempUnknownTopics = new HashSet<>();
-            topicsNotReady = validateTopics(topicsNotReady, topics, tempUnknownTopics);
-            newlyCreatedTopics.addAll(topicsNotReady);
-
+            final Set<NewTopic> validatedTopicObjects = createValidatedTopicObjects(topics, topicsNotReady, newlyCreatedTopics);
+            if (!validatedTopicObjects.isEmpty()) {
+                setupValidatedTopics(validatedTopicObjects, topicsNotReady);
+            }
             if (!topicsNotReady.isEmpty()) {
-                final Set<NewTopic> newTopics = new HashSet<>();
+                maybeThrowTimeoutExceptionDuringMakeReady(deadlineMs);
+            }
+        }
+        log.debug("Completed validating internal topics and created {}", newlyCreatedTopics);
 
-                for (final String topicName : topicsNotReady) {
-                    if (tempUnknownTopics.contains(topicName)) {
-                        // for the tempUnknownTopics, don't create topic for them
-                        // we'll check again later if remaining retries > 0
-                        continue;
-                    }
-                    final InternalTopicConfig internalTopicConfig = Objects.requireNonNull(topics.get(topicName));
-                    final Map<String, String> topicConfig = internalTopicConfig.properties(defaultTopicConfigs, windowChangeLogAdditionalRetention);
+        return newlyCreatedTopics;
+    }
+
+    private Set<NewTopic> createValidatedTopicObjects(final Map<String, InternalTopicConfig> topics,
+                                                            Set<String> topicsNotReady,
+                                                            final Set<String> newlyCreatedTopics) {
+        final Set<String> tempUnknownTopics = new HashSet<>();
+
+        topicsNotReady = validateTopics(topicsNotReady, topics, tempUnknownTopics);
+        newlyCreatedTopics.addAll(topicsNotReady);
+
+        final Set<NewTopic> validatedTopicObjects = new HashSet<>();
+
+        for (final String topicName : topicsNotReady) {
+            if (tempUnknownTopics.contains(topicName)) {
```

Comment:
If we change `validateTopics` (cf other comment above) to only return topic we want to create, we won't need this check.